salt/_renderers/better_kubernetes.py

```diff
@@ -0,0 +1,104 @@
+"""A renderer for Kubernetes YAML manifests.
+
+Given a Kubernetes YAML file (which may be a stream of objects, i.e. YAML
+snippets separated by `---` lines), this will render a sequence of states
+(represented as an OrderedDict), mapping every such object to an invocation
+of our custom `object_[present|absent]` state function.
+
+To use it, add a shebang like `#!kubernetes` as the first line of your
+manifests SLS file. Optionally, you can use rendering pipelines (if templating
+is required), e.g. `#!jinja | kubernetes`.
+
+The shebang also supports passing options to this renderer, in the format
+`#!kubernetes argument1=value1&argument2=value2` (basically, a query string in
+the `application/x-www-form-urlencoded` format).
+The supported options are:
+- `kubeconfig`, the path to the kubeconfig file to use for communicating with
+  K8s API
+- `context`, the context from the kubeconfig to use
+- `absent`, a boolean to toggle which state function variant (`object_present`
+  or `object_absent`) to use (defaults to False)
+
+TODO: improve management of `kubeconfig` and `context` parameters, relying on
+      master configuration and sane defaults - look into `__opts__`
+"""
+from salt.exceptions import SaltRenderError
+from salt.ext import six
+from salt.utils.yaml import SaltYamlSafeLoader
+from salt.utils.odict import OrderedDict
+import yaml
+
+__virtualname__ = 'kubernetes'
+
+
+def __virtual__():
+    return __virtualname__
+
+
+def _step_name(obj, absent=False):
+    try:
+        namespace = obj.metadata.namespace
```

Comment:
It could: `CustomObject`s don't have validation of the manifest except for `apiVersion` and `kind`. I could add it to the constructor though, to make it clearer. Or, for simplicity, just check `obj.metadata.name` first since it's mandatory, and then use `getattr(obj.metadata, 'namespace', None)` without fear.